### ISSUE TITLE
Optimize bundle size

### ISSUE THREAD

**Issue by vigneshshanmugam** [elastic/apm-agent-rum-js#159]
Some possible optimisations that we can refactor in our code base to reduce the bundle size. 

+  **Config object** contains lots of keys with long names that gets used across most of the files which cannot be mangled by the minifier resulting in bundle size increase
+ **Accessing properties of window**, We do a lof of checks for properties in the window and we have to check if a property/function exists in the window (Ex: window.performance - atleast checked more than 10 times), These checks cannot be mangled by the minifier and most of the libraries do something smart here by wrapping in an IIFE
```js
((function(perf){
  var entries = perf.getEntries();
  var timing = perf.timing

})(window.performance)
```

In the above case, window.performance is used in only one place and rest of the code are mangled. 
+ **Factory pattern for services**, do we even need them? They add some size trying to pass around and also for performance. 

+ **Unused Code**: Transaction and stack service has unused code that is required only for testing . and can be refactored. 

+ **ES6 Modules**: To utilise some of the performance techniques like Tree shaking with the bundlers/minifiers, We have to migrate the codebase to es6 modules.

**Comment by vigneshshanmugam** [comment]
### Savings: 

#### properties of window: 

I tried to experiment with fine tuning the window properties ( window.performance, window.location, window.fetch) as per the above mentioned example. 


Filename | Before | After 
-------- | --------| ---------
elastic-apm-js-base.min.js |  53521  | 53098
elastic-apm-js-base.min.js (After gzipping) | 16.3 kB | 16.1 kB

**Comment by vigneshshanmugam** [comment]
Savings: 

**Factory Pattern for services**

Filename | Before | After 
-------- | --------| ---------
elastic-apm-js-base.min.js |  54047  | **51913**
elastic-apm-js-base.min.js (After gzipping) | 16.04KB | **15.81KB**

**Comment by roncohen** [comment]
interesting! thanks for investigating. I wonder if you think there's added complexity when removing the factory pattern?

**Comment by vigneshshanmugam** [comment]
Final changes after removing factory pattern 

Filename | Before(master) | After 
-------- | --------| ---------
elastic-apm-js-base.min.js |  54047  | **51570**

**Comment by jahtalab** [comment]
We just had a very productive meeting, we decided to keep the service factory.

Some of the reasons to keep the service factory:
* We don't want to create file level instances which creates issues with our testing
* We have the option to initialise the services at init time instead of the load time. This is because this can increase the loading time of our script while the init can happen asynchronously and at the time the user decides to init the agent.
* We can have the option to disable some services, for example disable all of the error monitoring services if the user doesn't use error monitoring.

However, the optimisations in bundle size shown here can be done regardless of the service factory pattern so we decided to do those while still using the service factory.
